\hspace{24pt}

\section{Conclusion}
In order to reduce the influence of reference bias and improve the accuracy of EAGLE for evaluating variants, we have expanded the function of EAGLE.  While the original version of EAGLE relies on a standard read mapping (pile-up) to find reads relevant to a candidate variant, we use a read-index to find additional reads which match the variant genome sequence (but not necessarily match the reference genome sequence).  We modified EAGLE to allow us to add these reads to the pile-up before EAGLE evaluates the candidate variant.

After this series of steps, we conducted an experiment that simulates the real situation to verify our method. The experimental results of our simulation show that our method can effectively reduce the influence of reference bias, and for longer indels the effect is particularly significant. At the same time, our modified EAGLE we can work normally in any read coverage situation.

We also examined dbSNP, a real data set of candidate variants.  We observed good results as before, although the effect on SNPs calls is relatively limited.  Our implementation is well integrated into EAGLE and can reduce reference bias by missing fewer relevant reads.  Our use of a read-index requires only a modest time overhead.  The additional memory required is significant, but still manageable.

\section{Future Work}
This thesis focused on comparing the results of EAGLE before and after using a read-index to consider more reads.
As future work, our EAGLE with read-index method should also be compared to other variant candidate evaluation methods on
more comprehensive benchmark datasets.
